libyul/AsmParser.cpp

```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
 	return nullptr;
 }
 
+langutil::Token Parser::advance()
+{
+	auto const token = ParserBase::advance();
+	if (m_sourceMap)
+		fetchSourceLocationFromComment();
+	return token;
+}
+
+void Parser::fetchSourceLocationFromComment()
+{
+	solAssert(m_sourceMap.has_value(), "");
+
+	if (m_scanner->currentCommentLiteral().empty())
+		return;
+
+	static std::regex const lineRE = std::regex(
+		R"~~~(\s*@src\s+(\d+):(-1|\d+):(-1|\d+)\s*)~~~",
+		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	);
+
+	string_view text = m_scanner->currentCommentLiteral();
+	std::cmatch cm;
+	if (!std::regex_search(text.data(), text.data() + text.size(), cm, lineRE))
+		return;
+
+	try
+	{
+		solAssert(cm.size() == 4, "");
+		auto const sourceIndex = stoul(cm[1].str());
+		solAssert(sourceIndex <= std::numeric_limits<unsigned long>::max(), "");
+		auto const start = stoi(cm[2].str());
+		auto const end = stoi(cm[3].str());
+		solAssert(start <= end, "");
+
+		auto charStreamIt = m_sourceMap->find(static_cast<unsigned>(sourceIndex));
+		if (charStreamIt == m_sourceMap->end())
+		{
+			// Invalid source index is currently silently ignored.
+			// The SourceLocation from the Scanner will be used instead.
+			m_errorReporter.syntaxError(2674_error, m_scanner->currentLocation(), "Invalid source mapping. Source index not defined via @use-src.");
+			m_locationOverride.reset();
+			return;
+		}
+		shared_ptr<CharStream> charStream = charStreamIt->second;
+		solAssert(charStream, "");
+
+		m_locationOverride = SourceLocation{start, end, charStream};
+	}
+	catch (std::out_of_range const&)
+	{
+		m_errorReporter.syntaxError(6367_error, m_scanner->currentLocation(), "Integer value in source mapping out of range.");
+		m_locationOverride.reset();
+	}
+}
+
 Block Parser::parseBlock()
 {
 	RecursionGuard recursionGuard(*this);
 	Block block = createWithLocation<Block>();
 	expectToken(Token::LBrace);
 	while (currentToken() != Token::RBrace)
 		block.statements.emplace_back(parseStatement());
-	block.debugData = updateLocationEndFrom(block.debugData, currentLocation());
+	if (!m_locationOverride)
```

Comment:
I'm not sure this is the right switch. What happens if we have
```
{
/// @src ...
let x := 1
}
```

The value of `!m_locationOverride` changes while we are parsing the block.

Reply:
It might be that we need to distinguish three settings:
 - always take source location from scanner
 - always take source location from "location override" (and do not update it during parsing)
 - use empty source location at start (also not taking it from the scanner) and change it for every `@src`.

Reply:
we have an enum class for that now.